### PR TITLE
Introduce Tinkernet multisig XCM configs to Kusama/Rococo through xcm-builder

### PR DESCRIPTION
**Pull request by arrudagates**
This PR introduces the XCM configs required for Saturn multisigs from Tinkernet to transact on Kusama/Rococo, this is done simply by providing implementations to convert the MultiLocation to both an AccountId and a Signed RuntimeOrigin.

For more information regarding why this approach of using a very specific derivation function and defining this function in the receiver chain's runtime is necessary for Saturn, and also general information about Saturn, you can check the discussion that was opened in the forum prior: https://forum.polkadot.network/t/saturn-xcm-multisig-integration-on-kusama-a-technical-discussion/2694

For a more high level overview of Saturn, check this article: https://invarch.medium.com/saturn-the-future-of-multi-party-ownership-ac7190f86a7b

Please ask any questions left unanswered by the forum post and I'll make sure to clarify them!

~~Obs: My editor applied auto-format to some sections of the files edited, if those are unwanted for any reason, let me know and I'll revert them.~~
Those were actually the incorrect formatting standard required by CI, so I fixed them.

### CODE REVIEW

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/saturn-xcm-multisig-integration-on-kusama-a-technical-discussion/2694/2

File: runtime/kusama/src/xcm_config.rs
Comment (rphmeier) [inline runtime/kusama/src/xcm_config.rs:450]: needs docs

File: runtime/kusama/src/xcm_config.rs
Comment (xlc) [inline runtime/kusama/src/xcm_config.rs:443]: no overflow handling?

File: runtime/kusama/src/xcm_config.rs
Comment (xlc) [inline runtime/kusama/src/xcm_config.rs:440]: some comments to explain the magic numbers

File: runtime/kusama/src/xcm_config.rs
Comment (arrudagates) [reply runtime/kusama/src/xcm_config.rs:455]: Doc comments added!

File: runtime/kusama/src/xcm_config.rs
Comment (arrudagates) [reply runtime/kusama/src/xcm_config.rs]: While the id should technically never be over u32::MAX, you're right that it should be handled here as Kusama/Rococo shouldn't have to trust that to be true. Added overflow handling!

File: runtime/kusama/src/xcm_config.rs
Comment (arrudagates) [reply runtime/kusama/src/xcm_config.rs:443]: Comments added!

File: runtime/rococo/src/xcm_config.rs
Comment (KiChjang) [inline runtime/rococo/src/xcm_config.rs:405]: Instead of putting the logic here and repeating it on the XCM config on Kusama, let's put all of this under `xcm/xcm-builder` so that the logic is shared between the two (and upcoming) runtimes.

Comment (arrudagates) [reply runtime/rococo/src/xcm_config.rs:405]: Will do!

Comment (arrudagates) [reply runtime/rococo/src/xcm_config.rs]: Done!

File: xcm/xcm-builder/Cargo.toml
Comment (KiChjang) [inline xcm/xcm-builder/Cargo.toml:50]: Use tabs.

Comment (arrudagates) [reply xcm/xcm-builder/Cargo.toml]: 😅

File: xcm/xcm-builder/src/derivation.rs
Comment (KiChjang) [inline xcm/xcm-builder/src/derivation.rs:1]: It makes more sense to me that this should be put in `location_conversion.rs`, since this function is doing exactly that.

Comment (arrudagates) [reply xcm/xcm-builder/src/derivation.rs:1]: Comment didn't apply to a code section, are you referring to derivation.rs?

File: xcm/xcm-builder/src/location_conversion.rs
Comment (KiChjang) [inline xcm/xcm-builder/src/location_conversion.rs:290]: Put these hardcoded literals into constants, so that the comments are unnecessary here.

Comment (arrudagates) [reply xcm/xcm-builder/src/location_conversion.rs:290]: Should I declare the constants above `pub fn derive_tinkernet_multisig` or on the top of the file?

Comment (arrudagates) [reply xcm/xcm-builder/src/location_conversion.rs]: I declared them on the top, let me know if you'd rather I change it.

**Review by KiChjang — APPROVED** [review]

**Comment by arrudagates** [comment]
@xlc @rphmeier Could you review this PR again?

**Comment by arrudagates** [comment]
@KiChjang Could you add review requests to other maintainers in this PR?

**Comment by arrudagates** [comment]
@ggwpez @rphmeier @xlc Sorry for the ping again, but this PR has been waiting for further approvals for quite a while, could you review it or forward this request to others?

**Review by rphmeier — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2956824

**Review by sam0x17 — APPROVED** [review]

File: xcm/xcm-builder/src/location_conversion.rs
Comment (ggwpez) [inline xcm/xcm-builder/src/location_conversion.rs:372]: Is this `expect` strictly necessary?  
Otherwise I would link "[When to use .expect()](https://substrate.stackexchange.com/questions/317/when-to-use-expect)" and rather just return an error.

**Review by ggwpez — APPROVED** [review]

File: xcm/xcm-builder/src/location_conversion.rs
Comment (kianenigma) [inline xcm/xcm-builder/src/location_conversion.rs:350]: This test is assuming your kusama para-id, right? it should be in the kusama runtime file, not in XCM-builder.

File: xcm/xcm-builder/src/origin_conversion.rs
Comment (kianenigma) [inline xcm/xcm-builder/src/origin_conversion.rs:249]: Similarly, I think from an organization perspective, it is more correct for this configuration to be placed in the kusama runtime, as it is the only user of it. But I see that it is more "convenient" to do it like this.

File: xcm/xcm-builder/src/location_conversion.rs
Comment (arrudagates) [reply xcm/xcm-builder/src/location_conversion.rs:372]: This is written based on the derivation function on pallet-multisig: https://github.com/paritytech/substrate/blob/master/frame/multisig/src/lib.rs#L507

File: xcm/xcm-builder/src/origin_conversion.rs
Comment (arrudagates) [reply xcm/xcm-builder/src/origin_conversion.rs:249]: Kusama is not the only user of these converters, parachains will also import these into their xcm configurations from xcm-builder. I believe this also responds the previous comment.

Comment (ggwpez) [reply xcm/xcm-builder/src/origin_conversion.rs:249]: I think for this converter it is fine. But about the previous comment: the `TINKERNET_PARA_ID` is only valid on Kusama, right?

File: xcm/xcm-builder/src/location_conversion.rs
Comment (arrudagates) [reply xcm/xcm-builder/src/location_conversion.rs]: Replaced the expect with an error.

File: xcm/xcm-builder/src/location_conversion.rs
Comment (arrudagates) [reply xcm/xcm-builder/src/location_conversion.rs]: Renamed the constants to include the KUSAMA_ prefix.

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
bot rebase

**Comment by ggwpez** [comment]
bot merge

**Comment by arrudagates** [comment]
🎉

**Comment by gavofyork** [comment]
Please ensure that items specific to Kusama which go into common crates like `xcm-builder` are:

a) placed in a Kusama-specific path or have "Kusama" clearly placed in its name; and
b) convert the `MultiLocation` value to be universal and then that used for matching.

This code makes little sense outside of the very specific context it is designed for. In particular, matching code like `parents: _` is not tight enough, especially when combined with a non-consensus-specific `Parachain` junction matcher. First converting into a universal location and then matching on that would alleviate these issues.

**Comment by arrudagates** [comment]
> b) convert the `MultiLocation` value to be universal and then that used for matching.

@gavofyork From what I understand, to do this I need to send a `UniversalOrigin` instruction with a `GlobalConsensus` junction, which will then mutate the origin similarly to how `DescendOrigin` works. But then I hit an immediate blocker, as this instruction is currently unsupported in Kusama/Rococo (which was very hard to figure out, since the way it's made unsupported is by giving it max weight and no meaningful logs): https://github.com/paritytech/polkadot/blob/master/runtime/kusama/src/weights/xcm/mod.rs#L242

**Pull request merged**
